File: 程式設計：資料結構答案檔/Go/401.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func readGraphFromFile(filename string) ([][]int, int, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, 0, err
	}
	defer file.Close()

	var graph [][]int
	scanner := bufio.NewScanner(file)
	numVertices := 0

	for scanner.Scan() {
		line := scanner.Text()
		tokens := strings.Split(line, ",")
		if numVertices == 0 {
			numVertices = len(tokens)
		}
		row := make([]int, numVertices)
		for i, token := range tokens {
			value, err := strconv.Atoi(token)
			if err != nil {
				return nil, 0, err
			}
			row[i] = value
		}
		graph = append(graph, row)
	}
	return graph, numVertices, nil
}

func main() {
	graph, numVertices, err := readGraphFromFile("read.txt")
	if err != nil {
		fmt.Println("Error reading graph from file:", err)
		return
	}

	degrees := make([]int, numVertices)
	var maxWeight, minWeight int = 0, 1<<31 - 1

	for i := 0; i < numVertices; i++ {
		weightSum := 0
		for j := 0; j < numVertices; j++ {
			if graph[i][j] > 0 {
				degrees[i]++
				weightSum += graph[i][j]
			}
		}
		if weightSum > maxWeight {
			maxWeight = weightSum
		}
		if weightSum < minWeight {
			minWeight = weightSum
		}
	}

	fmt.Print("Degree:")
	for i, deg := range degrees {
		fmt.Print(deg)
		if i < numVertices-1 {
			fmt.Print(",")
		}
	}
	fmt.Println()
	fmt.Println(maxWeight - minWeight)
}

File: 程式設計：資料結構答案檔/C++/401.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <cstdlib>
#include <cstring>

#define MAX_VERTICES 60

int graph[MAX_VERTICES][MAX_VERTICES];
int num_vertices = 0;

void readGraphFromFile(const std::string& filename) {
    std::ifstream file(filename);
    if (!file) {
        std::cout << "File could not be opened.\n";
        exit(EXIT_FAILURE);
    }

    std::string line;
    int row = 0;
    while (getline(file, line)) {
        std::istringstream iss(line);
        std::string token;
        int col = 0;
        while (getline(iss, token, ',')) {
            graph[row][col++] = std::stoi(token);
        }
        if (row == 0) num_vertices = col;
        row++;
    }
    file.close();
}

int main() {
    readGraphFromFile("read.txt");

    int degrees[MAX_VERTICES] = {0};
    int max_weight = 0, min_weight = 1000000;

    for (int i = 0; i < num_vertices; i++) {
        int weight_sum = 0;
        for (int j = 0; j < num_vertices; j++) {
            if (graph[i][j] > 0) {
                degrees[i]++;
                weight_sum += graph[i][j];
            }
        }
        if (weight_sum > max_weight) max_weight = weight_sum;
        if (weight_sum < min_weight) min_weight = weight_sum;
    }

    std::cout << "Degree:";
    for (int i = 0; i < num_vertices; i++) {
        std::cout << degrees[i];
        if (i < num_vertices - 1) std::cout << ",";
    }
    std::cout << std::endl << max_weight - min_weight << std::endl;

    return 0;
}
